File: HW7/131044075/BigramDyn.cpp
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
//
//#include "BigramDyn.h"


#include "BigramDyn.h"
#include <fstream>
#include <iostream>
#include <map>

// CONSTRUCTORS

template <class T>
BigramDyn<T>::BigramDyn(int dataType) :capacity(50), used(0)
{
	dType = dataType;
    data = new pair< pair<T,T>, int >[capacity];
}

template <class T>
BigramDyn<T>::BigramDyn(const BigramDyn& myProg) :capacity(myProg.capacity), used(myProg.used)
{
	dType = myProg.dataType;
    data = new pair< pair<T,T>, int >[capacity];
    for(int i=0; i<capacity; i++)
        data[i] = myProg.data[i];
}

// = operator overload
template <class T>
BigramDyn<T>& BigramDyn<T>::operator =(const BigramDyn& myProg){
    
    if(capacity != myProg.capacity)
    {
        delete [] data;
        data = new string[myProg.capacity];
    }

    capacity = myProg.capacity;
    used = myProg.used;
    for(int i=0; i<used; i++)
        data[i] = myProg.data[i];
    return *this;
}

//helper function for << overload
template<class T>
void BigramDyn<T> :: getBigrams(ostream& outputStream)
{

    int max = data[0].second;
    
    for(int i=0; data[i].second != '\0'; ++i)
    {
        if(data[i].second > max)
        {
            max=data[i].second;
            ++i;
        }
    }
    
    for(int i=0; data[i].second != '\0'; ++i)
    {
        for(int j=0; data[j].second != '\0'; ++j)
        {
            if(data[j].second == max)
            {
                outputStream << data[j].first.first << " " << data[j].first.second
                        << " = " << data[j].second << endl;
            }
        }
        max--;
    }
}

// getter private data "int used"
template <class T>
int BigramDyn<T>::getUsed(){
    return used;
}

//getter for private data "data"
template <class T>
pair< pair<T,T>, int>* BigramDyn<T>::getData(){
    return data;
}

//takes a filename as a string parameter, reads the file, calculates all the bigrams.
template <class T>
void BigramDyn<T>::readFile(string filename){
    
    string temp, tmp;
    int foundItemIndex = -1;
	int count=0;
    ifstream infile, inCount;
    infile.open(filename);
	inCount.open(filename);
    if(infile.fail()){
        throw myExceptions("File reading error!");
    }   

	
	while(inCount>>tmp)
			count++;

	if(count==0)
			throw myExceptions("Empty Document");

	inCount.close();

    getline(infile, temp);

	if(dType==1)
    {
    
    int convertString;
   
        for(int j=0; temp[j]!='\0' ; ++j)
        {
                convertString = temp[j]-'0';
                if(!((convertString>=0 && convertString<=9 ) || temp[j]==' ') )
                        throw myExceptions("Wrong data type");			
        }	
         
    }

    
    if(dType == 3)
    {
        int convertString;
        for(int i=0; temp[i]!='\0' ; ++i){
            convertString = temp[i]-'0';
                if(!((convertString>=0 && convertString<=9 )|| temp[i]=='.' || temp[i]==' '))
                        throw myExceptions("Wrong data type");
        }
    }






    
    istringstream iss(temp);
    T word1, word2;
    iss >> word1;
    while(iss >> word2) {

        pair <T,T> icPair (word1,word2);
        
        pair< pair<T,T>, int> disPair (icPair, 0);
        
        for(int i=0; i<used; i++){
            if(icPair == data[i].first)
                foundItemIndex = i;
        }
            
        //Bulunamadıysa
        if(foundItemIndex == -1){
            if(used < capacity){
                data[used] = disPair;
                used++;
            }
            else{
                capacity *=2;
                pair< pair<T,T>, int >* tempData = new pair< pair<T,T>, int >[capacity];

                for(int i=0; i<used; i++)
                    tempData[i] = data[i];

                delete [] data;
                data = new pair< pair<T,T>, int >[capacity];

                for(int i=0; i<used; i++)
                    data[i] = tempData[i];

                delete [] tempData;

                data[used] = disPair;
                used++;
            }
            data[used-1].second = 1;
        }
        //Bulunduysa
        else {
            data[foundItemIndex].second++;
        }
        word1 = word2;
        foundItemIndex = -1;
    }


    infile.close();

}


//adds data to pair
template <class T>
void BigramDyn<T>::pushBack(pair< pair<T,T>, int > instruction){
    
    if(used < capacity){
        data[used] = instruction;
        used++;
    }
    else{
        capacity *=2;
        pair< pair<T,T>, int >* tempData = new pair< pair<T,T>, int >[capacity];

        for(int i=0; i<used; i++)
            tempData[i] = data[i];

        delete [] data;
        data = new pair< pair<T,T>, int >[capacity];

        for(int i=0; i<used; i++)
            data[i] = tempData[i];

        delete [] tempData;

        data[used] = instruction;
        used++;
    }
}


//DESTRUCTOR
template <class T>
BigramDyn<T>::~BigramDyn(){
    
    delete [] data;
}

template <class T>
int BigramDyn<T>::numGrams(){
    
    int total = 0;
    
    for(int i=0; i<used; i++){
        total += data[i].second;
    }

    return total;

}

//takes two elements as bigrams and returns the number of that bigram read so far
template <class T>
int BigramDyn<T>::numOfGrams(T first, T second){
    
    pair <T,T> icPair(first, second);
    
    for(int i=0; i<used; i++){
        if(icPair == data[i].first)
            return data[i].second;
    }
    
    return 0;
}

//returns the bigram that has occurred most frequently
template <class T>
pair<T,T> BigramDyn<T>::maxGrams(){
    
    
    int tempMax = 0;
    
    int lastIndex;
    
    for(int i=0; i<used; i++){
        if(data[i].second > tempMax){
            tempMax = data[i].second;
            lastIndex = i;
        }
    }
    
    return data[lastIndex].first;
}

File: HW8/131044075/BigramDyn.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.lang.reflect.Array;
import java.util.Scanner;
import java.util.StringTokenizer;

/**
 * Created by sude on 3.1.2017.
 */
public class BigramDyn<T> implements Bigram<T>{

    private int dType;
    private int capacity;
    private int used;
    private int count = 0;

    private T[] data;

    public BigramDyn(int dataType) {
        dType = dataType;
        capacity = 50;
        used = 0;
        data = (T[]) new Object[capacity];
    }

    public int getdType() {
        return dType;
    }

    public int getCapacity() {
        return capacity;
    }

    public int getUsed() {
        return used;
    }

    public T[] getData() {
        return data;
    }

    public void setdType(int dType) {
        this.dType = dType;
    }

    public void setCapacity(int capacity) {
        this.capacity = capacity;
    }

    @Override
    public String toString() {

        StringBuilder stringBuilder = new StringBuilder();
        String[] array = new String[used];
        String tempStr = null;

        for(int i=1; i<used; i++){

            String bigramElement = numOfGrams(data[i-1],data[i]) + " : (" +  data[i-1].toString() + " , " + data[i].toString() + ")" + "\n";
            array[i-1] = bigramElement;
            //stringBuilder.append(array[i-1]);
        }

        for (int i=0; i<used-1; i++)
        {
            for (int j=1; j<used-i-1; j++)
            {
                if(array[j-1].compareTo(array[j]) < 0)
                {
                    tempStr = array [j-1];
                    array [j-1] = array [j];
                    array [j] = tempStr;
                }
            }
        }

        for(int i=1; i<used-1; i++){
            if(!array[i].equals(array[i-1]))
                stringBuilder.append(array[i-1]);
        }
        stringBuilder.append(array[used-2]);


        return stringBuilder.toString();




//        //int max = data[0].second;
//
//        for(int i=0; data[i].second != '\0'; ++i)
//        {
//            if(data[i].second > max)
//            {
//                max=data[i].second;
//                ++i;
//            }
//        }
//
//        for(int i=0; data[i].second != '\0'; ++i)
//        {
//            for(int j=0; data[j].second != '\0'; ++j)
//            {
//                if(data[j].second == max)
//                {
//                    outputStream << data[j].first.first << " " << data[j].first.second
//                            << " = " << data[j].second << endl;
//                }
//            }
//            max--;
//        }
    }

    @Override
    public void readFile(String filename) throws myExceptions, IOException {

        String temp, tmp;
        int foundItemIndex = -1;
        int count=0;

        BufferedReader br = null;
        br = new BufferedReader(new FileReader(filename));
        temp = br.readLine();
        br.close();

        StringTokenizer st = new StringTokenizer(temp);
        count = st.countTokens();

        if(count==0)
            throw new myExceptions("Exception: Empty Document");

        if(dType==1)
        {

            int convertString;
            for(int j=0; j<temp.length() ; ++j)
            {
                convertString = temp.charAt(j)-'0';
                if(!((convertString>=0 && convertString<=9 ) || temp.charAt(j)==' ') )
                    throw new myExceptions("Wrong data type");
            }

        }


        if(dType == 3)
        {
            int convertString;
            for(int i=0; i<temp.length() ; ++i){
                convertString = temp.charAt(i)-'0';
                if(!((convertString>=0 && convertString<=9 )|| temp.charAt(i)=='.' || temp.charAt(i)==' '))
                    throw new myExceptions("Wrong data type");
            }
        }

        String[] arr = temp.split(" ");
        String word1;



        for (int i=0; i<arr.length; i++) {

            if(used<capacity) {
                data[used] = (T) arr[i];
                used++;
            }
            else{
                capacity *= 2;
                T[] tempData = (T[]) new Object[capacity];

                for(int j=0; j<used; j++)
                    tempData[j] = data[j];
                data = null;
                data = (T[]) new Object[capacity];

                for(int j=0; j<used; j++)
                    data[j] = tempData[j];

                data[used] = (T) arr[i];
                tempData = null;
                used++;

            }
        }

    }

    @Override
    public int numGrams() {
        return used-1;
    }

    @Override
    public int numOfGrams(T firstElement, T secondElement) {

        int count = 0;
        T word1, word2;
        word1 = data[0];
        for(int i=1; i<used; i++){

            word2 = data[i];
            if(firstElement.toString().equals(word1.toString()) && secondElement.toString().equals(word2.toString())){
                count++;
            }
            word1 = word2;
        }

        return count;
    }
}
